service/handlers.go

```diff
@@ -111,67 +104,66 @@ func SignHandler(w http.ResponseWriter, r *http.Request) {
 
 	if r.Body == nil {
 		w.WriteHeader(http.StatusBadRequest)
-		formatSignResponse(false, "error-nil-data", "", "Uninitialized POST data", "", w)
+		formatSignResponse(false, "error-nil-data", "", "Uninitialized POST data", nil, w)
 		return
 	}
 
-	assertions := new(Assertions)
-	err := json.NewDecoder(r.Body).Decode(&assertions)
+	// Read the full request body
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		panic(err)
+	}
+	if len(data) == 0 {
+		w.WriteHeader(http.StatusBadRequest)
+		formatSignResponse(false, "error-sign-empty", "", "No data supplied for signing", nil, w)
+		return
+	}
 
 	defer r.Body.Close()
 
-	switch {
-	// Check we have some data
-	case err == io.EOF:
-		w.WriteHeader(http.StatusBadRequest)
-		formatSignResponse(false, "error-sign-empty", "", "No data supplied for signing", "", w)
-		return
-		// Check for parsing errors
-	case err != nil:
+	// Use the ubuntu-core assertions module to decode the body and validate
+	assertion, err := asserts.Decode(data)
+	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
 		errorMessage := fmt.Sprintf("%v", err)
-		formatSignResponse(false, "error-decode-json", "", errorMessage, "", w)
+		formatSignResponse(false, "error-decode-json", "", errorMessage, nil, w)
 		return
 	}
 
-	// Validate the model by checking that it exists on the database
-	model, err := Environ.DB.FindModel(assertions.Brand, assertions.Model, assertions.Revision)
-	if err != nil {
-		w.WriteHeader(http.StatusNotFound)
-		formatSignResponse(false, "error-model-not-found", "", "Cannot find model with the matching brand, model and revision", "", w)
+	// Check that we have a device-serial assertion (the details will have been validated by Decode call)
+	if assertion.Header("type") != "device-serial" {
```

Comment:
We could use the Assertion interface method "Type()" and use the type registry "AssertionType" instead of the header. That way, this won't break if they decide to modify the header.